### PR TITLE
graph/{topo,community}: add clique graph and k-clique community finding

### PR DESCRIPTION
**Pull request by kortschak**
@mewmew @vladimir-ch Please take a look.

There are likely optimisations in `community.KCliqueCommunities` since we are using `topo.CliqueGraph` which does more work than is absolutely required, but at least for the moment the clarity in the code is worth that additional cost, which is probably significantly less than the unavoidable costs of doing the maximal clique finding in the first place. If it becomes an issue for someone, we can re-address it then.

### CODE REVIEW

**Comment by mewmew** [comment]
@kortschak Just to let you know, I'm in my first week at Uni now, and will have a bit less time on my hands. Cheers /u

**Comment by kortschak** [comment]
No worries, they're always just invitations.

File: graph/community/k_communities.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	"gonum.org/v1/gonum/graph/internal/set"
+	"gonum.org/v1/gonum/graph/simple"
+	"gonum.org/v1/gonum/graph/topo"
+	"gonum.org/v1/gonum/graph/traverse"
+)
+
+// KCliqueCommunities returns the k-clique communties of the undirected graph g for
+// k greater than zero. The returned communities are identified by linkage via k-clique
+// adjacency, where adjacency is defined as having k-1 common nodes. KCliqueCommunities
+// returns a single component including/ the full set of nodes of g when k is 1,
```
Comment (vladimir-ch) [inline graph/community/k_communities.go]: Extra `/`.

File: graph/community/k_communities_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
```
Comment (vladimir-ch) [inline graph/community/k_communities_test.go]: 2015?

File: graph/topo/clique_graph.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+
+	for _, cliques := range cliqueNodes {
+		for i, uc := range cliques {
+			for _, vc := range cliques[i+1:] {
+				// Retain the nodes that contribute to the
+				// edge between the cliques.
+				var edgeNodes []graph.Node
+				switch 1 {
```
Comment (vladimir-ch) [inline graph/topo/clique_graph.go]: The first time I see this construct, wow.

File: graph/community/k_communities.go (hunk 1 above)
Comment (kortschak) [reply graph/community/k_communities.go]: Done.

File: graph/community/k_communities_test.go (hunk 2 above)
Comment (kortschak) [reply graph/community/k_communities_test.go]: Done.

**Comment by kortschak** [comment]
Fixes pushed, please re-approve.

File: graph/topo/clique_graph.go (hunk 3 above)
Comment (kortschak) [reply graph/topo/clique_graph.go]: Yes, `switch {...` is just sugar for `switch true {...`. There are interesting an clever ways that you can abuse this.

**Pull request merged**